### ISSUE TITLE
To handle TCP broken pipe errors

### ISSUE THREAD

**Issue by bmmcq** [frankmcsherry/timely-dataflow#185]
Hi,
I have run a long-drawn "timely" job in the cluster mode, and there were TCP broken pipe errors sometimes, but the "timely" job didn't know that and blocked.  As a user, can "timely" notify me with error messages and close the job when errors occur in the network, instead of blocked endlessly?

**Comment by frankmcsherry** [comment]
It is meant to. This sounds like a similar issue to #184. A timely computation should abort if workers send to a broken tcp connection, and a heartbeat is one way to make sure that happens.

**Comment by bmmcq** [comment]
I get another way. In fact, TIMELY's communication threads can recognize TCP errors when they try to send or receive data through the network. 
 
In the TIMELY codes : 

```rust 
    fn send_loop(&mut self) {
        let mut stash = Vec::new();
        while let Ok((header, buffer)) = self.sources.recv() {
            // ...  
            for (header, mut buffer) in stash.drain_temp() {
                // ... ... 
                header.write_to(&mut self.writer).unwrap(); // <-- unwrap() panic send thread when TCP connection is broken.
                self.writer.write_all(&buffer[..]).unwrap(); // as above;
                buffer.clear();
            }

            self.writer.flush().unwrap();    // <-- as above
        }
    }
```
The "send-thread" is panic and exits without notify anyone else when TCP connections break. 
It's nearly same in the "recv-thread". 

So, if the communication threads can notify the network abnormal event, for example, with an atomic boolean, the worker threads can check this and abort current computation immediately. 

For example : 

```rust
    // The NETWORK_ABNORMAL is an atomic boolean indicating error occurred;
    pub fn step(&mut self) -> bool {
        (!NETWORK_ABNORMAL.load(Ordering::SeqCst)) && {
            let mut active = false;
            for dataflow in self.dataflows.borrow_mut().iter_mut() {
                let sub_active = dataflow.step();
                active = active || sub_active;
            }

            // discard completed dataflows.
            self.dataflows.borrow_mut().retain(|dataflow| dataflow.active());
            active
        }
    }
```

I have tested this solution and it seems working. The overhead of "AtomicBool#load" is negligible when there is no error in the network in my benchmark.

There may be something I don't take into account. Looking forward to your suggestions.

**Comment by frankmcsherry** [comment]
This *should* be fixed with changes made in #135. If it is still happening, then there is a bug and we should work on reproducing it!
